### PR TITLE
integration with quantized gmm kernel

### PR DESCRIPTION
**Pull request by Bob-Chen222**
## Motivation
Integrate a new quantized GEMM kernel and support quantization for both MoE and standard layers.

## Modifications
This change refactors the quantization logic. Because qwix only supports weights used in `einsum` and `dot_product`, MoE kernel weights must be quantized manually. I added `quantize_tensor`, `dequantize_tensor`, and related utilities to implement a workaround enabling w8a8 quantization for MoE models. I also added support for custom quantization policies via a config file (e.g., int8.yaml) using a new `quantization_config` data class, and added `fp8_grok.yaml`, aligned with Grok-2 quantization in sglang-gpu.

## Accuracy Tests
I have tested grok-2 on gpqa for a few times and seems like that the performance is not decreased. I can have a rigorous test that test 8 or 16 times and get the average difference between quantized and unquantized version.

## Benchmarking and Profiling
The benchmarking and profiling

image
image

The quantized version (left) currently provides only a minimal throughput improvement over the unquantized version (right), indicating room for further optimization. I plan to profile this PR tomorrow. The implementation is functionally correct, and the overall design structure is in place.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [x] (Optional) The necessary documentation update.

### CODE REVIEW

File: python/sgl_jax/srt/layers/moe.py [hunk 1]
```diff
@@ -6,14 +6,18 @@
...
+)
 
 
 class GateLogit(nnx.Module):
     def __init__(
         self,
         input_size: int,
         num_experts: int = 0,
-        weight_dtype: jnp.dtype = jnp.bfloat16,
+        weight_dtype: jnp.dtype = jnp.float32,
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/moe.py]: default type changed

File: python/sgl_jax/srt/layers/moe.py [hunk 2]
```diff
@@ -36,7 +40,7 @@ def __init__(
             self.bias = None
 
     def __call__(self, hidden_states: jax.Array) -> tuple[jax.Array, jax.Array | None]:
-        logits = hidden_states.astype(self.weight_dtype) @ self.kernel
+        logits = hidden_states.astype(jnp.float32) @ self.kernel
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/moe.py:44]: Is this change made to achieve higher accuracy?

File: python/sgl_jax/srt/utils/quantization/configs/fp8_w8a8.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# FP8 weight and activation quantization config
+# Works for both dense and MoE models
+
+quantization:
+  # Qwix rules for dense layers (attention, MLP, embeddings)
+  qwix:
+    rules:
+      - module_path: '.*'
+        weight_qtype: 'float8_e4m3fn'
+        act_qtype: 'float8_e4m3fn'
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/utils/quantization/configs/fp8_w8a8.yaml:10]: MoE-related layers were not excluded.

Comment (Bob-Chen222) [reply python/sgl_jax/srt/utils/quantization/configs/fp8_w8a8.yaml:10]: This is a bit tricky. Native qwix only supports quantization for weights/activations used in three operator types: einsum, dot_general, and conv_general_dilate. As a result, the GEMM (Pallas op) is not quantized, and the moe section in the YAML specifically targets this GEMM path.

However, within the overall epmoe module, einsum is still used for the unpermute step and dot_general is used for the gating operation. These parts therefore continue to rely on qwix for quantization.

File: python/sgl_jax/srt/layers/moe.py (hunk 2 above)
Comment (Bob-Chen222) [reply python/sgl_jax/srt/layers/moe.py:44]: This behavior matches what I observed in sglang-GPU, which is why I made the second change to explicitly cast to float32. I will revert the first change since it has no impact on Grok.

After retesting 10 times, I found that using bf16 vs. float32 for activations has a negligible effect that is difficult to measure, so I’m not sure converting activation to fp32 is actually necessary. That said, we could still apply it to align the implementation, as long as it does not negatively affect other models.
Screenshot 2026-01-09 at 12 02 29 PM

File: python/sgl_jax/srt/utils/quantization/configs/fp8_w8a8.yaml (hunk 3 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/utils/quantization/configs/fp8_w8a8.yaml:10]: So this part is for quantizing the gate, right?

Comment (Bob-Chen222) [reply python/sgl_jax/srt/utils/quantization/configs/fp8_w8a8.yaml:10]: Yes the quantization of gate will be handled by qwix. And in the grok-2 specific setting I excluded the quantization of gate

**Comment by JamesBrianD** [comment]
@Bob-Chen222 Since fused MoE was recently merged, could you help rebase the branch and test the baseline performance of Qwen3-MoE for both quantized and non-quantized setups? Compare with this baseline https://github.com/sgl-project/sglang-jax/pull/349
